types/context.go

```diff
@@ -189,7 +189,9 @@ func (c Context) WithBlockTime(newTime time.Time) Context {
 }
 
 func (c Context) WithBlockHeight(height int64) Context {
-	return c.withValue(contextKeyBlockHeight, height)
+	newHeader := c.BlockHeader()
+	newHeader.Height = height
+	return c.withValue(contextKeyBlockHeight, height).withValue(contextKeyBlockHeader, newHeader)
```

Comment:
Can we get rid of the duplication? Maybe we should remove the header from the context altogether and just store its components.

Reply:
OK